### PR TITLE
feat: fp8 block scaling

### PR DESCRIPTION
**Pull request by jiemingz**
# What does this PR do ?

**Add a one line overview of what this PR aims to accomplish.**

# Issues
List issues that this PR closes ([syntax](https://docs.github.com/en/issues/tracking-your-work-with-issues/using-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)):

# Usage
* **You can potentially add a usage example below**

```python
# Add a code snippet demonstrating how to use this 
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](/NVIDIA/NeMo-RL/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you run the unit tests and functional tests locally? Visit our [Testing Guide](/NVIDIA/NeMo-RL/blob/main/docs/testing.md) for how to run tests
- [ ] Did you add or update any necessary documentation? Visit our [Document Development Guide](/NVIDIA/NeMo-RL/blob/main/docs/documentation.md) for how to write, build and test the docs.

# Additional Information
* ...

### CODE REVIEW

File: nemo_rl/models/generation/fp8.py [hunk 1]
```diff
@@ -0,0 +1,415 @@
...
+
+    if USE_WEIGHT_POW2_SCALE:
+        # Calculate exponent HW instruction: cvt.rp.satfinite.ue8m0x2.f32
+        exponent = torch.ceil(torch.log2(descale))
+        # Post process exponent to be in range of -127 to 127 and to be E8M0 biased
+        exponent = torch.clamp(exponent, min=-127, max=127) + 127
+        # Convert to uint8 container
+        exponent = exponent.to(torch.uint8)
+        # Calculate descale_fp to apply to data_hp
+        descale_fp = torch.where(
```
Comment (vcuinv) [inline nemo_rl/models/generation/fp8.py]: I feel starting from here` descale_fp` should be called `scale_fp` and `descale_fp_inv`  should be called `descale_fp`. From what I have seen if `scale` and `descale` both appear, `scale` means the factor applied on the high precision data and `descale` means the factor passed to the quantized gemm kernel.

File: nemo_rl/models/generation/fp8.py [hunk 2]
```diff
@@ -0,0 +1,415 @@
...
+    )
+    layer.weight_scale_inv = _create_param_from_subclass_attributes(
+        BlockQuantScaleParameter(
+            data=weight_scale_inv, output_dim=0, input_dim=1,
+            weight_loader=layer.weight_scale_inv.weight_loader)
+    )
+
+
+@triton.jit
+def _per_token_group_quant_fp8(
```
Comment (vcuinv) [inline nemo_rl/models/generation/fp8.py:442]: "per_token_group" in name gives me a feeling that the quantization granularity covers a group of tokens, but actually it is sub_token?

File: nemo_rl/models/generation/fp8.py [hunk 3]
```diff
@@ -0,0 +1,415 @@
...
+            1.0,
+            torch.exp2(127 - exponent.to(torch.float32)),
+        )
+        descale_fp_inv = torch.reciprocal(descale_fp)
+    else:
+        descale_fp = max_dtype / max_abs
+        descale_fp = torch.where(
+            max_abs == 0, 1.0, descale_fp
+        )  # preserve the behavior for 0 amax case
+        descale_fp = torch.where(
```
Comment (vcuinv) [inline nemo_rl/models/generation/fp8.py]: Should value of `inf` also be changes to `1.0` instead of `torch.finfo(data_hp.dtype).max`?

File: nemo_rl/models/generation/fp8.py (hunk 2 above)
Comment (jiemingz) [reply nemo_rl/models/generation/fp8.py:442]: yeah, it is unfortunately named, but we maybe should keep the function name because it patches a vllm function;  I can add a comment
https://github.com/vllm-project/vllm/blob/7f280d69c98e560427d2cbc9c3c3c13a83510dca/vllm/model_executor/layers/quantization/utils/fp8_utils.py#L243

File: nemo_rl/models/generation/fp8.py (hunk 3 above)
Comment (jiemingz) [reply nemo_rl/models/generation/fp8.py]: do you mean changing this to?
```
        descale_fp = torch.where(
            max_abs == torch.inf, 1.0, descale_fp
        ) 
```

Comment (vcuinv) [reply nemo_rl/models/generation/fp8.py]: Yep.

**Review by rybakov — DISMISSED** [review]
Should we also add a config, RL/examples/configs/grpo_math_8B_fp8_L3_F1_G_i.yaml 
For example, below config can be a good candidate (with optionally set       num_last_layers_in_bf16: 0      num_first_layers_in_bf16: 0):


# GRPO Algorithm Configuration
defaults: "grpo_math_1B.yaml"

grpo:
  num_prompts_per_step: 64
  num_generations_per_prompt: 32

loss_fn:
  use_importance_sampling_correction: true

policy:
  model_name: "meta-llama/Llama-3.1-8B-Instruct"
  tokenizer:
    name: ${policy.model_name} ## specify if you'd like to use a tokenizer different from the model's default
  train_global_batch_size: 512
  train_micro_batch_size: 1
  generation_batch_size: 32 # Only used when generating using HF backend
  logprob_batch_size: 2
  max_total_sequence_length: 4096
  precision: "bfloat16"
  fsdp_offload_enabled: false
  activation_checkpointing_enabled: false

  dtensor_cfg:
    enabled: True

  dynamic_batching:
    train_mb_tokens: 4096
    logprob_mb_tokens: 8192

  optimizer:
    name: "torch.optim.AdamW"
    kwargs:
      lr: 3.0e-7
      weight_decay: 0.01
      betas: [0.9, 0.999]
      eps: 1e-8

  scheduler:
    - name: "torch.optim.lr_scheduler.LinearLR"
      kwargs:
        start_factor: 0.1
        end_factor: 1.0
        # The scheduler iteration is per GPRO step and is decoupled with the optimizer step (may be >=1 per GPRO step)
        total_iters: 13
    - name: "torch.optim.lr_scheduler.ConstantLR"
      kwargs:
        factor: 1.0
        total_iters: 10000000000
    - milestones: [13]

  generation:
    backend: "vllm"
    max_new_tokens: ${policy.max_total_sequence_length}
    temperature: 1.0
    top_p: 1.0
    top_k: null
    stop_token_ids: null
    stop_strings: null
    vllm_cfg:
      precision: 'fp8'
      use_deep_gemm: true
      num_last_layers_in_bf16: 3
      num_first_layers_in_bf16: 1
      tensor_parallel_size: 1
      gpu_memory_utilization: 0.6
      max_model_len: ${policy.max_total_sequence_length}

cluster:
  gpus_per_node: 8
  num_nodes: 1

File: nemo_rl/models/generation/vllm_backend.py [hunk 4]
```diff
@@ -25,50 +25,6 @@
         "If you are working interactively, you can install by running  `uv sync --extra vllm` anywhere in the repo."
     )
 
-
-def _patch_gemma3_mm():
```
Comment (SahilJain314) [inline nemo_rl/models/generation/vllm_backend.py]: Any particular reason this is removed?

File: nemo_rl/models/generation/vllm_backend.py [hunk 5]
```diff
@@ -131,7 +87,8 @@ def update_weights_from_local_ipc_handles(self, local_device_ipc_handles):
         Returns:
             bool: True if weights were successfully updated.
         """
-        try:
+        # try:
```
Comment (SahilJain314) [inline nemo_rl/models/generation/vllm_backend.py]: Assuming this is debug?

File: nemo_rl/models/generation/fp8.py [hunk 6]
```diff
@@ -0,0 +1,489 @@
+import torch, os
```
Comment (SahilJain314) [inline nemo_rl/models/generation/fp8.py]: Would really appreciate an associated doc with this explaining the various quantization schemes and configurable items here.

File: nemo_rl/algorithms/grpo.py [hunk 7]
```diff
@@ -309,6 +309,11 @@ def setup(
         )
     elif backend == "vllm":
         generation_config = cast(VllmConfig, generation_config)
+        if generation_config["vllm_cfg"]["precision"] == "fp8":
+            assert loss_config["use_importance_sampling_correction"] is True, (
+                "Importance sampling must be enabled for vLLM FP8 generation!"
```
Comment (SahilJain314) [inline nemo_rl/algorithms/grpo.py]:
```suggestion
                "Importance sampling must be enabled for vLLM FP8 generation for good convergence!"
```

File: pyproject.toml [hunk 8]
```diff
@@ -55,7 +55,9 @@ automodel = [
     "flash-attn==2.7.4.post1",
 ]
 vllm = [
-    "vllm==0.9.0",
+    "cuda-python",
```
Comment (SahilJain314) [inline pyproject.toml:66]: I see that the dependencies are updated here but the uv.lock is unchanged. Is that actually the case for you locally? If not, can you commit the updated uv.lock?

File: nemo_rl/models/generation/vllm_backend.py (hunk 4 above)
Comment (jiemingz) [reply nemo_rl/models/generation/vllm_backend.py]: Im following this comment:
https://github.com/NVIDIA-NeMo/RL/blob/1da0fdaf4280e3975775d7b9776f915f7441e608/nemo_rl/models/generation/vllm_backend.py#L38

**Comment by jiemingz** [comment]
image

File: docs/fp8.md [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+                num_first_layers_in_bf16: 1
+                # Use FP32 scaling factors. Rounding scaling factors to the nearest pow2 may improve quantization 
+                # fidelity however this feature is still under research.
+                use_weight_pow2_scale: False
+                use_activation_pow2_scale: False
+```
+
+## Accuracy
+
+We observe on the Llama 8b recipe a ~5% accuracy loss is incurred with FP8 generations. Convergence is still under active research and FP8 generations should be used with caution. We are investigating ways to close the accuracy gap and further improve performance.
```
Comment (parthchadha) [inline docs/fp8.md:39]: @jiemingz I think we should add reference llama8b plots to this doc.

File: nemo_rl/models/generation/fp8.py [hunk 10]
```diff
@@ -0,0 +1,558 @@
...
+            param.__class__ = param.subclass_type
+    # Finally load the weights into vllm
+    model.load_weights(weights_quantized)
+    # Undo the type change above to the original type
+    for name, param in model.named_parameters():
+        if hasattr(param, "subclass_type"):
+            param.__class__ = param.orig_type
+
+
+def kitchen_block_scale(
```
Comment (parthchadha) [inline nemo_rl/models/generation/fp8.py]: @jiemingz do we need to call kitchen*? This is standard block scaling right?

Comment (jiemingz) [reply nemo_rl/models/generation/fp8.py]: no; yes-the function is only lifted from kitchen, i'll rename it for clarity

File: docs/fp8.md [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+                num_first_layers_in_bf16: 1
+                # Use FP32 scaling factors. Rounding scaling factors to the nearest pow2 may improve quantization 
+                # fidelity however this feature is still under research.
+                use_weight_pow2_scale: False
+                use_activation_pow2_scale: False
+```
+
+## Accuracy
+
+We observe on the Llama 8b recipe a ~2% accuracy loss is incurred with FP8 generations. Convergence is still under active research and FP8 generations should be used with caution. We are investigating ways to close the accuracy gap and further improve performance.
```
Comment (SahilJain314) [inline docs/fp8.md]:
```suggestion
We observe on the Llama 8b recipe a ~2% accuracy loss is incurred with FP8 generations. Convergence is still under active research and FP8 generations should be used with caution. We are investigating ways to close the accuracy gap and further improve performance. Note: This gap appears after ~2 *thousand* steps of RL. Prior to that, similar to other works, we find a minimal gap in accuracy
```

**Review by parthchadha — DISMISSED** [review]
